### PR TITLE
gh-103615: Use local events for opcode tracing

### LINKED ISSUE #108982: test_sys_settrace tests are not independent
Running any of TraceOpcodesTestCase tests from test_sys_settrace always fails. For example:
```
$ ./python -m test -v test_sys_settrace -m test.test_sys_settrace.TraceOpcodesTestCase.test_01_basic
...
test_01_basic (test.test_sys_settrace.TraceOpcodesTestCase.test_01_basic) ... FAIL

======================================================================
FAIL: test_01_basic (test.test_sys_settrace.TraceOpcodesTestCase.test_01_basic)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/serhiy/py/cpython/Lib/test/test_sys_settrace.py", line 388, in test_01_basic
    self.run_test(basic)
  File "/home/serhiy/py/cpython/Lib/test/test_sys_settrace.py", line 364, in run_test
    self.run_and_compare(func, func.events)
  File "/home/serhiy/py/cpython/Lib/test/test_sys_settrace.py", line 360, in run_and_compare
    self.compare_events(func.__code__.co_firstlineno,
  File "/home/serhiy/py/cpython/Lib/test/test_sys_settrace.py", line 1745, in compare_events
    self.assertLess(len(skip_opcode_events), len(events),
AssertionError: 3 not less than 3 : No 'opcode' events received by the tracer
...
```

Running several of TraceOpcodesTestCase tests only fails the first of them.

Running also `RaisingTraceFuncTestCase.test_line_event_raises_before_opcode_event` makes the tests passing successfully.
```
$ ./python -m test -v test_sys_settrace -m test.test_sys_settrace.TraceOpcodesTestCase.test_01_basic -m test_line_event_raises_before_opcode_event
...
test_line_event_raises_before_opcode_event (test.test_sys_settrace.RaisingTraceFuncTestCase.test_line_event_raises_before_opcode_event) ... ok
test_01_basic (test.test_sys_settrace.TraceOpcodesTestCase.test_01_basic) ... ok
...
```

### PR DESCRIPTION
**Pull request by gaogaotiantian**
The opcode tracing of PEP 669 to support legacy tracing is not ideal. I talked about it in #103615 but I'll list the gist here:
1. The user has to set `f_trace_opcodes` BEFORE `sys.settrace()`, which is different than what it used to do (I'd consider this as a bug actually).
2. Once `f_trace_opcodes` is set once, you set it globally and forever. The instruction event callback will ALWAYS trigger on every single instruction when you are tracing, even if `f_trace_opcodes` is set to `False` on that frame. It will stick even if you turn off the trace and back on. (there will be no event because in the callback, `f_trace_opcodes` is explicitly checked, but it's a performance hit).

The best way to fix this is to use local events for opcode tracing. By nature, opcode tracing is enabled frame by frame, there's no way to enable it globally for every code object from Python, so that's what the underlying structure should do.

There are 4 occasions to enable instrumentation on instruction event on a code object:
1. when we do `frame.f_trace_opcode = True` if trace is enabled already
2. when we enter a frame on a CALL event - the instruction instrumentation could be stripped because we turned off trace on this frame, but `frame.f_trace_opcode` may persist.
3. when we enable trace with `frame.f_trace_opcode == True`
4. when we set `frame.f_trace = True` - this is to trace the current frame before a call.

There is only 1 case when we need to disable the instrumentation:
1. when the instruction event fires and we realize `f_trace_opcode == False`
2. However, just to make the obvious case, I also added the disabling when we set `f_trace_opcode = False`

With this mechanism, we eliminate the global `f_opcode_trace_set`.

<details>
<summary>
The performance improvement is significant on cases when opcode events is only needed from a small piece of code:
</summary>

```python
import sys
import time

def g():
    a = 0
    for _ in range(100000):
        a += 1
    return a

def f():
    for _ in range(10):
        g()

events = 0

def trace(frame, event, arg):
    global events
    if event == 'opcode':
        events += 1
    return trace

frame = sys._getframe()
frame.f_trace_opcodes = True
sys.settrace(trace)
start = time.time()
f()
print(time.time() - start)
sys.settrace(None)
print(events)
```

For the code above, the time saving is about 50%.

</details>

Oh, this also fixes #108982.

* Issue: gh-103615

### CODE REVIEW

**Comment by markshannon** [comment]
Looks good.
I like the design, it looks like the performance impact will be low and it definitely looks better than what we have now.

It could do with some tests beyond the implicit one in https://github.com/python/cpython/issues/108982.
Can you turn https://github.com/python/cpython/issues/103615#issue-1673607411 into a test?

**Comment by gaogaotiantian** [comment]
> Looks good. I like the design, it looks like the performance impact will be low and it definitely looks better than what we have now.
> 
> It could do with some tests beyond the implicit one in #108982. Can you turn [#103615 (comment)](https://github.com/python/cpython/issues/103615#issue-1673607411) into a test?

The one in #103615 is a bit tricky - it will only reproduce if `f_trace_opcode` is never set before. I can do it in a separate process and collect the output. Do you think that's worth the effort?

**Comment by markshannon** [comment]
I think it is OK to add as a normal test without using subprocesses.
It is a regression test. It should never fail in the future, it doesn't really matter if it only failed erratically in the past.

Maybe add it to the monitoring tests, rather than sys_settrace tests? The way it will fail if only test.test_monitoring is run.

**Comment by gaogaotiantian** [comment]
> I think it is OK to add as a normal test without using subprocesses. It is a regression test. It should never fail in the future, it doesn't really matter if it only failed erratically in the past.
> 
> Maybe add it to the monitoring tests, rather than sys_settrace tests? That way it will fail if only test.test_monitoring is run.

Well if the regression test does not fail on the old code, it does not help that much. If we only run a normal test, then it's doing exactly the same thing as our existing tests for `f_trace_opcodes`. Putting it in `test_monitoring` works, but I don't think that's a very decent solution.

I created the working regression test for this specific case. It needs some extra libraries as it requires a new file to execute in a separate process. However, it's the test we need for this scenario. Let me know if you have suggestions on the test.

**Comment by gaogaotiantian** [comment]
Hi @markshannon , just fixed the conflict, is there anything I should update on this PR? Thanks!

File: Python/legacy_tracing.c [hunk 1]
```diff
@@ -107,6 +107,33 @@ sys_profile_call_or_return(
...
+_PyEval_SetOpcodeTrace(
+    PyFrameObject *frame,
+    bool enable
+) {
+    assert(frame != NULL);
+    assert(PyCode_Check(frame->f_frame->f_executable));
+
+    PyCodeObject *code = (PyCodeObject *)frame->f_frame->f_executable;
+    _PyMonitoringEventSet events = 0;
+    if (_PyMonitoring_GetLocalEvents(code, PY_MONITORING_SYS_TRACE_ID, &events) == 0) {
```
Comment (markshannon) [inline Python/legacy_tracing.c]: Minor style issue.
We generally check for errors, not success, so we can early exit.
```C
   if (_PyMonitoring_GetLocalEvents(code, PY_MONITORING_SYS_TRACE_ID, &events) < 0) {
        return -1;
   }
   /* Rest of code goes here */
```

**Review by markshannon — COMMENTED** [review]
One style issue, otherwise looks good.

File: Python/legacy_tracing.c (hunk 1 above)
Comment (gaogaotiantian) [reply Python/legacy_tracing.c]: Fixed, thanks!

**Comment by markshannon** [comment]
Thanks for doing this

**Pull request merged**